devices/r4isdhc.cpp

```diff
@@ -171,32 +171,92 @@ bool writeNor(const uint32_t dest_address, const uint32_t length, const uint8_t
     return true;
 }
 
-bool trySecureInit(BlowfishKey key) {
-    union {
-        uint32_t u32;
-        uint8_t bytes[4];
-    } buf;
+bool sendRawInit() {
+    if (ntrcard::state.status != ntrcard::Status::RAW) {
+        logMessage(LOG_ERR, "r4isdhc: sendRawInit: status (%d) not RAW",
+            static_cast<uint32_t>(ntrcard::state.status));
+        return false;
+    }
+
+    CmdBuf4 buf;
+    // this is actually the NOR write disable command
+    // the r4isdhc will respond to cart commands with 0xFFFFFFFF if
+    // the "magic" command hasn't been sent, so we check for that
+    sendCommand(0x40199, 4, buf.u8, 0x180000);
+    if (buf.u32 != 0xFFFFFFFF) {
+        logMessage(LOG_ERR, "r4isdhc: sendRawInit: pre-test returned 0x%08X", buf.u32);
+        return false;
+    }
+
+    if (platform::CAN_RESET) {
+        // actually if we're here CAN_RESET is true
+        // but.. can't be too safe
+        ntrcard::init();
+    }
+    sendCommand(0x68, 4, nullptr, 0x180000);
+    // now it will return zeroes
+    sendCommand(0x40199, 4, buf.u8, 0x180000);
+    if (buf.u32 == 0) {
+        return true;
+    }
+
+    logMessage(LOG_ERR, "r4isdhc: sendRawInit: post-test returned 0x%08X", buf.u32);
+    return false;
+}
+
+bool sendKey2Init() {
+    if (ntrcard::state.status != ntrcard::Status::KEY2) {
+        logMessage(LOG_ERR, "r4isdhc: sendKey2Init: status (%d) not KEY2",
+            static_cast<uint32_t>(ntrcard::state.status));
+        return false;
+    }
+
+    CmdBuf4 buf;
+    sendCommand(0x66, 4, nullptr, 0x586000);
+    ntrcard::state.status = ntrcard::Status::RAW;
+
+    sendCommand(0x40199, 4, buf.u8, 0x180000);
 
-    ntrcard::init();
+    // FIXME this is a really poor check
+    // a non-r4isdhc cart will stay in KEY2 and likely return something that isn't all-FF
+    if (buf.u32 != 0xFFFFFFFF) {
+        return true;
+    }
+
+    logMessage(LOG_ERR, "r4isdhc: sendKey2Init: post-test returned 0x%08X", buf.u32);
+    ntrcard::state.status = ntrcard::Status::KEY2;
+    return false;
+}
+
+bool trySecureInit(BlowfishKey key) {
+    if (ntrcard::state.status != ntrcard::Status::RAW) {
+        if (platform::CAN_RESET) {
+            ntrcard::init();
```

Comment:
it looks can move `CAN_RESET` check to `ntrcard::init`.
then need to check initialize result.

Reply:
`ntrcard::init` _already_ checks `platform::CAN_RESET`, and bails if it is false.

Reply:
so why not use `trySecureInit` that result?

Reply:
Huh?

Reply:
why not use `if (ntrcard:init())`, and it can replace `platform::CAN_RESET` test